### PR TITLE
Add ShellCommand config option and change default to /bin/sh (on Unix)

### PR DESCRIPTION
**Pull request by xonixx**
Attempt at fixing #59.

I checked the test suite (`go test ./... -v`) - seems all green.
1. Though I feel would be good to add some more tests for this specific case, but I'm a bit unsure how exactly. The test suite in `interp_test.go` doen't have means for provinding environment variables to test. Would be glad for an advice.
2. Hopefully I changed `sh` -> `/bin/sh` in all relevant places (to cover `system()`, `"cmd" | getline`, `print | "cmd"`). But now it feels like a bit of copy-paste. Do you think it makes sense to do something about this - like extracting a function or a constant? If yes, my concern is - what is better place for it, since the change is in two files, but it doesn't feel right to complicate things much for little reason.

### CODE REVIEW

**Comment by benhoyt** [comment]
Thanks for this. It's a good start, though I think we should make a few improvements:

* As mentioned in #59, while we're at it I think we should make this configurable via `interp.Config`. Let's add a new field `ExecCommand []string`, which if nil, defaults to `[]string{"/bin/sh", "-c"}` on Unix, and `[]string{"sh", "-c"}` on Windows.
* Setting the Windows default to just "sh" without the /bin path will make it keep working as it does today on Windows. See the unit test failures on Windows: https://github.com/benhoyt/goawk/pull/64/checks?check_run_id=3550794167
* I'm open to other names for this config field too. `ExecArgs`?
* With the new field, we will need a couple of additional tests for it. Probably a new `TestExecCommand` function in `interp/interp_test.go`. See for example https://github.com/benhoyt/goawk/blob/0eb3f8c8b7889fe1feff86aed767dcdb30cbc98e/interp/interp_test.go#L1014

**Comment by xonixx** [comment]
First things first I'm totaly agree with your arguments and I'm ready to do the changes you request.

However please find couple of considerations below.

Regarding failing on Windows. 
Just did a quick check https://github.com/xonixx/playground/runs/3568823431?check_suite_focus=true#step:3:8
I see that on Win this uses sh.exe bundled with Git, which is pre-bundled into Windows environment on Github (https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md). 
Therefore I assume on clean windows (no Git) goawk will fail on `system()` as well.

Does it make sense to specify in docs that on Windows the `system()` functionality of goawk is only available when Git or msys (or other sh.exe provider) is installed?

Also does it make sense to not be Windows-specific but rather go more generic approach and implement the adaptive logic to attempt both
- `/bin/sh` as POSIX require
- `sh` from `$PATH` if above fails

@benhoyt if you think it doesn't make sense to go this deep I'm totally fine just to do as you already prosposed.

**Comment by benhoyt** [comment]
It's a good point regarding "sh" only existing on Windows due to Git. However, I'd rather not do a system call in every ExecProgram call due to the cost of hitting the file system (I know some use that in a fairly tight loop). So for now let's do what I suggested, and I can clean up the Windows shell 
handling in a subsequent commit (maybe "cmd.exe").

File: interp/functions.go [hunk 1]
```diff
@@ -261,6 +261,15 @@ func (p *interp) callBuiltin(op Token, argExprs []Expr) (value, error) {
 	}
 }
 
+// executes code using configured system shell
+func (p *interp) execShell(code string) *exec.Cmd {
+	shellInterpreter := p.shellCommand[0]
+	shellArgs := p.shellCommand[1:]
```
Comment (benhoyt) [inline interp/functions.go:267]: Naming nit: how about we name these `executable` and `args`? We don't need the `shell` prefix in this context as this function already includes that in the name.

File: interp/functions.go [hunk 2]
```diff
@@ -320,11 +329,11 @@ func (p *interp) callUser(index int, args []Expr) (value, error) {
 	return null(), nil
 }
 
-// Call native-defined function with given name and arguments, return
+// Call native-defined function with given name and arguments,
```
Comment (benhoyt) [inline interp/functions.go:332]: Thanks for catching. Let's make this "return its return value" like the comment on `callUser`.

File: interp/functions.go [hunk 3]
```diff
@@ -320,11 +329,11 @@ func (p *interp) callUser(index int, args []Expr) (value, error) {
...
 // return value (or null value if it doesn't return anything).
 func (p *interp) callNative(index int, args []Expr) (value, error) {
 	f := p.nativeFuncs[index]
-	minIn := len(f.in) // Mininum number of args we should pass
+	minIn := len(f.in) // Minimum number of args we should pass
```
Comment (benhoyt) [inline interp/functions.go:336]: Good catch!

File: interp/interp.go [hunk 4]
```diff
@@ -247,6 +252,19 @@ func ExecProgram(program *Program, config *Config) (int, error) {
 		}
 	}
 
+	// Setup system shell command
+	if len(config.ShellCommand) != 0 {
+		p.shellCommand = config.ShellCommand
+	} else {
+		var shellInterpreter string
```
Comment (benhoyt) [inline interp/interp.go:259]: Nits: let's call this `executable`. Also, let's simplify slightly to:

```go
executable := "/bin/sh"
if runtime.GOOS == "windows" {
    executable = "sh"
}
```

File: interp/interp_test.go [hunk 5]
```diff
@@ -1029,6 +1030,25 @@ func TestConfigVarsCorrect(t *testing.T) {
...
+			})
+		testGoAWK(t, `BEGIN { "world" | getline; print }`, "", "hello world\n", "", nil,
+			func(config *interp.Config) {
+				config.ShellCommand = []string{"/bin/echo", "hello"}
+			})
+		testGoAWK(t, `BEGIN { print "hello world" | "-" }`, "", "hello world\n", "", nil,
+			func(config *interp.Config) {
+				config.ShellCommand = []string{"/bin/cat"}
+			})
+	}
```
Comment (benhoyt) [inline interp/interp_test.go:1049]: Nice, thanks for adding a test!

For a Windows test, it [looks like](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/cmd) we might be able to test with `config.ShellCommand = []string{"cmd.exe", "/c"}` ... though I'm not quite sure what would be a good, Windows-only command to run. Feel free to leave that out.

**Review by benhoyt — CHANGES_REQUESTED** [review]
Looks good, thanks! Just a couple of style nits.

**Comment by xonixx** [comment]
@benhoyt Hopefully I fixed all CR notes

**Comment by benhoyt** [comment]
Thanks, looks good!

**Pull request merged**